language/move-lang/src/naming/uses.rs

```diff
@@ -50,63 +61,110 @@ impl<'a> Context<'a> {
             return;
         }
+        let (upstream, downstream) = match dep_type {
+            DepType::Use => (current_mident, mident),
+            DepType::Friend => (mident, current_mident),
+        };
 
         let m = self
             .neighbors
-            .entry(self.current_module.clone().unwrap())
+            .entry(upstream)
+            .or_insert_with(BTreeMap::new)
+            .entry(downstream)
             .or_insert_with(BTreeMap::new);
-        if m.contains_key(uses) {
+        if m.contains_key(&dep_type) {
             return;
         }
+        m.insert(dep_type, loc);
+    }
 
-        m.insert(uses.clone(), loc);
+    fn add_usage(&mut self, mident: ModuleIdent, loc: Loc) {
+        self.add_neighbor(mident, DepType::Use, loc);
     }
 
-    fn dependency_graph(&self) -> DiGraphMap<&ModuleIdent, ()> {
-        let edges = self
-            .neighbors
-            .iter()
-            .flat_map(|(parent, children)| children.iter().map(move |(child, _)| (parent, child)));
-        DiGraphMap::from_edges(edges)
+    fn add_friend(&mut self, mident: ModuleIdent, loc: Loc) {
+        self.add_neighbor(mident, DepType::Friend, loc);
     }
 }
 
-fn cycle_error(context: &Context, cycle_ident: ModuleIdent) -> Error {
-    let cycle = shortest_cycle(&context.dependency_graph(), &cycle_ident);
-
-    // For printing uses, sort the cycle by location (earliest first)
-    let cycle_strings = cycle
+fn dependency_graph(
+    deps: &BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
+) -> DiGraphMap<&ModuleIdent, ()> {
+    let edges = deps
         .iter()
-        .map(|m| format!("'{}'", m))
-        .collect::<Vec<_>>()
-        .join(" uses ");
+        .flat_map(|(parent, children)| children.iter().map(move |(child, _)| (parent, child)));
+    DiGraphMap::from_edges(edges)
+}
 
-    let (used_loc, user, used) = best_cycle_loc(context, cycle);
+fn cycle_error(
+    deps: &BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
+    cycle_ident: ModuleIdent,
+) -> Error {
+    let graph = dependency_graph(deps);
+    let cycle = shortest_cycle(&graph, &cycle_ident);
+
+    let mut cycle_strings: String = cycle
+        .windows(2)
+        .map(|pair| {
+            let upstream = pair[0];
+            let downstream = pair[1];
+            let relations = deps.get(upstream).unwrap().get(downstream).unwrap();
+            let verb = if relations.contains_key(&DepType::Use) {
+                "uses"
+            } else {
+                assert!(relations.contains_key(&DepType::Friend));
+                "may use"
+            };
+            format!("'{}' {} ", upstream, verb)
+        })
+        .collect();
+    cycle_strings.push_str(&format!("'{}'", cycle.last().unwrap()));
 
-    let use_msg = format!("Invalid use of module '{}' in module '{}'.", used, user);
+    // For printing uses, sort the cycle by location (earliest first)
```

Comment:
I think the "friend" vs. "use" errors need to be a bit more different from each other. The "use" phrasing is fine the way it was, but "friend" is not a verb (at least IMO). Instead of "Invalid friend of module...", I suggest dropping the "of" and making it: "Invalid friend module...." We could even go one step further and drop the "module" since it should be clear that it is a module, e.g.: "Invalid friend '0x2::B' in module '0x2::A'"

Reply:
Thanks for the suggestion! I think "Invalid friend '0x2::B' in module '0x2::A'" sounds good to me.